### PR TITLE
feat: make DocList an actual Python List

### PR DESCRIPTION
**Pull request by JoanFM**
This PR needs to make sure that `DocList` is a List, thus fixing the issue that I cannot call `schema`  of a `BaseDoc` having nested `DocList`.

### CODE REVIEW

File: docarray/array/doc_list/sequence_indexing_mixin.py [hunk 1]
```diff
@@ -34,7 +37,7 @@ def _is_np_int(item: Any) -> bool:
     return False  # this is unreachable, but mypy wants it
 
 
-class IndexingSequenceMixin(Iterable[T_item]):
+class IndexingSequenceMixin(List[T_item]):
     """
     This mixin allow sto extend a list into an object that can be indexed
```
Comment (jupyterjazz) [inline docarray/array/doc_list/sequence_indexing_mixin.py:42]:
```suggestion
    This mixin allows to extend a list into an object that can be indexed
```

File: docarray/array/doc_list/sequence_indexing_mixin.py [hunk 2]
```diff
@@ -11,8 +11,11 @@
     cast,
     no_type_check,
     overload,
+    List,
 )
 
+from typing_extensions import SupportsIndex
```
Comment (samsja) [inline docarray/array/doc_list/sequence_indexing_mixin.py:17]: what is this SupportsIndex ?

File: docarray/array/doc_list/sequence_indexing_mixin.py [hunk 3]
```diff
@@ -139,7 +142,7 @@ def _del_from_indices(self: T, item: Iterable[int]) -> None:
             # each delete
             del self._data[ix]
```
Comment (samsja) [inline docarray/array/doc_list/sequence_indexing_mixin.py:143]: we still have `_data` ?

**Review by samsja — CHANGES_REQUESTED** [review]
but here we still have the `_data` things this looks weird no ?

**Comment by jupyterjazz** [comment]
this line is the problem
https://github.com/docarray/docarray/blob/8df9e7f3ae700ee84f766d38fd4cc2fb63908567/docarray/index/abstract.py#L424
after this change, it's always true

we could do this
```python
if not isinstance(docs, DocList) and isinstance(docs, List):
```

File: docarray/index/abstract.py [hunk 4]
```diff
@@ -892,7 +892,7 @@ def _convert_dict_to_doc(
                 doc_dict[field_name] = self._convert_dict_to_doc(inner_dict, t_)
 
         schema_cls = cast(Type[BaseDoc], schema)
-        return schema_cls(**doc_dict)
+        return schema_cls(**dict(doc_dict))
```
Comment (jupyterjazz) [inline docarray/index/abstract.py:895]: we should revert this change

File: docarray/array/doc_list/sequence_indexing_mixin.py (hunk 2 above)
Comment (jupyterjazz) [reply docarray/array/doc_list/sequence_indexing_mixin.py:17]: I think it's a fancy alternative for int

File: docarray/index/abstract.py (hunk 4 above)
Comment (samsja) [reply docarray/index/abstract.py]: done

File: docarray/array/doc_list/doc_list.py [hunk 5]
```diff
@@ -285,3 +287,11 @@ def __getitem__(self: T, item: IndexIterType) -> T:
 
     def __getitem__(self, item):
         return super().__getitem__(item)
+
+    @classmethod
+    def __class_getitem__(cls, item: Union[Type[BaseDoc], TypeVar, str]):
```
[the remark this answers is not in the document]
Comment (JoanFM) [reply docarray/array/doc_list/doc_list.py:292]: I do not quite understand this need? what `super` has another `_class_getitem`?

**Review by samsja — APPROVED** [review]

File: docarray/array/doc_list/doc_list.py [hunk 6]
```diff
@@ -194,31 +173,28 @@ def extend(self, docs: Iterable[T_doc]):
...
-        :param field: name of the fields to extract
-        :return: Returns a list of the field value for each document
-        in the doc_list like container
+        """Return all v  @classmethod
+          def __class_getitem__(cls, item: Union[Type[BaseDoc], TypeVar, str]):alues of the fields from all docs this doc_list contains
+        @classmethod
+          def __class_getitem__(cls, item: Union[Type[BaseDoc], TypeVar, str]):
+              :param field: name of the fields to extract
+              :return: Returns a list of the field value for each document
+              in the doc_list like container
```
Comment (jupyterjazz) [inline docarray/array/doc_list/doc_list.py:197]: note for myself to correct this docstring in my fastapi PR

**Review by jupyterjazz — COMMENTED** [review]
Nice! I'll try out fastapi integration now
